Apiseeds Lyrics plugin

avalloneandrea opened the pull request.

avalloneandrea committed:
Initial commit

phw commented on plugins/apiseeds-lyrics/apiseeds-lyrics.py:
```diff
@@ -0,0 +1,116 @@
+        data = load_json(response)
+        lyrics = data['result']['track']['text']
+        metadata['lyrics'] = lyrics
+        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
```
I think this and most other output here should actually be `log.debug`

phw commented on plugins/apiseeds-lyrics/apiseeds-lyrics.py:
```diff
@@ -0,0 +1,116 @@
+def process_result(album, metadata, response, reply, error):
+
+    try:
+        data = load_json(response)
```
Any specific reason why you use `load_json` here? Most often the better approach is to call `webservice.get()` with `parse_response_type='json'`.

If parsing fails `error` will be set. Also you should check for `error` first. If it is not empty log it and return early.

phw commented on plugins/apiseeds-lyrics/apiseeds-lyrics.py:
```diff
@@ -0,0 +1,116 @@
+def process_track(album, metadata, release, track):
+
+    apikey = config.setting['apiseeds_api_key']
+    if (apikey is None):
```
I think this is better written as `if not apikey:`. Also this check currently failed in my tests. I did not debug fully, but it likely was set to an empty string in my case and not `None`. But just check with `not apikey` and it's save and looks a bit more pythonic.

phw commented on plugins/apiseeds-lyrics/apiseeds-lyrics.py:
```diff
@@ -0,0 +1,116 @@
+        return
+
+    title = metadata['title']
+    if (title is None):
```
Same here, check for `if not artist` and `if not title`

phw commented on plugins/apiseeds-lyrics/apiseeds-lyrics.py:
```diff
@@ -0,0 +1,116 @@
+        APISEEDS_PORT,
+        apiseeds_path,
+        partial(process_result, album, metadata),
+        parse_response_type=None,
```
See my comments above, I think it would be better to set `parse_response_type='json'` here.

phw commented on plugins/apiseeds-lyrics/apiseeds-lyrics.py:
```diff
@@ -0,0 +1,116 @@
+        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
+
+    except:
+        log.info('{}: lyrics NOT found for track {}'.format(PLUGIN_NAME, metadata['title']))
```
Since this is logging in case of an exception I think you should log it as an error with `exc_info=True` (will include stack trace information) set. Something like:

```
log.error('{}: failed fetching lyrics for track {}'.format(PLUGIN_NAME, metadata['title']), exc_info=True)
```

Actually you can also do something like the following, but your use of `string.format` would be consistent with the rest of your code:

```
log.error('%s: failed fetching lyrics for track %s', PLUGIN_NAME, metadata['title'], exc_info=True)
```

avalloneandrea replied:
Actually, after replacing `load_json` with `parse_response_type='json'`, the only possibile exception is a KeyError raised by `data['result']['track']['text']`. I don't think it's useful to have the stack trace in this case.

phw replied:
Agreed, yes

phw reviewed (changes requested):
Thanks a lot for this, this is awesome. Pretty clean implementation. I have a couple of fixes and improvements, but actually nothing big. Looking forward to getting this merged.

avalloneandrea committed:
Small fixes

avalloneandrea committed:
Replace log.info and log.error with log.debug

avalloneandrea committed:
Replace load_json(response) with parse_response_type='json'

avalloneandrea committed:
Fix checks on apikey, artist and title

avalloneandrea committed:
Version 1.0.1

avalloneandrea committed:
Move processor to class

avalloneandrea committed:
Version 1.0.2

avalloneandrea committed:
Rename plugin to conform snake_case naming style

avalloneandrea commented:
I've implemented your suggestions in version 1.0.1, and moved the processor into it's own class in version 1.0.2...but maybe this wasn't a good idea, because it raised a _Method could be a function_ issue

phw commented:
Thanks for the update

> because it raised a _Method could be a function_ issue

You could just use `@staticmethod` on this method.

phw commented on plugins/apiseeds_lyrics/apiseeds_lyrics.py:
```diff
@@ -0,0 +1,122 @@
+            priority=True,
+            queryargs=apiseeds_params)
+
+    def process_response(self, album, metadata, document, reply, error):
```
You still need to check for `error` with something like:

```
if error:
    album._requests -= 1
    album._finalize_loading(error)
    return
```

phw commented on plugins/apiseeds_lyrics/apiseeds_lyrics.py:
```diff
@@ -0,0 +1,122 @@
+            metadata['lyrics'] = lyrics
+            log.debug('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
+
+        except:
```
Better check explicitly for `KeyError` that's the only one that should be raised here

avalloneandrea replied:
Added this:
```
if error:
    log.info('{}: lyrics NOT found for track {}'.format(PLUGIN_NAME, metadata['title']))
    album._requests -= 1
    album._finalize_loading(error)
    return
```

and this:
```
except KeyError:
```

but Picard crashes when the API is unable to find lyrics. This is the last thing I see in the console:
```
E: 11:24:18,237 webservice._handle_reply:383: Network request error for https://orion.apiseeds.com:443/api/music/lyric/artist/track?apikey=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
: Error transferring https://orion.apiseeds.com:443/api/music/lyric/artist/track?apikey=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 - server replied: Not Found (QT code 203, HTTP code 404)
```
How can I investigate the issue?

phw replied:
You probably should change it back to catch all exceptions again, but actually add the stacktrace option. This way you will also see what caused your crashes.

avalloneandrea replied:
I have double checked my code, this is working as expected now:
```
@staticmethod
def process_response(album, metadata, document, reply, error):

    if error:
        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
        album._requests -= 1
        album._finalize_loading(None)
        return

    try:
        lyrics = document['result']['track']['text']
        metadata['lyrics'] = lyrics
        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))

    except KeyError:
        log.info('{}: lyrics NOT found for track {}'.format(PLUGIN_NAME, metadata['title']))

    finally:
        album._requests -= 1
        album._finalize_loading(None)
```

However I notice that those three lines of code are written twice, and I'm not really a fan of code duplications:
```
log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
album._requests -= 1
album._finalize_loading(None)
```

Therefore I'm more prone to _"ask forgiveness not permission"_: don't check `error`, let `document` be `None`, and catch the `TypeError` raised by the first access to `document`:
```
@staticmethod
def process_response(album, metadata, document, reply, error):

    try:
        lyrics = document['result']['track']['text']
        metadata['lyrics'] = lyrics
        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))

    except (TypeError, KeyError):
        log.info('{}: lyrics NOT found for track {}'.format(PLUGIN_NAME, metadata['title']))

    finally:
        album._requests -= 1
        album._finalize_loading(None)
```

What do you think about it?

phw replied:
I think it should be fine. In case of `error != None` this error will be logged anyway also by the web service module.

It also depends a bit on what we expect this plugin to cause in case of an error. My original suggestion used `album._finalize_loading(error)`. This will cause the entire album loading to fail. Probably with a plugin like this we don't want to be so strict. So I'm totally fine with your solution.

phw commented on plugins/apiseeds_lyrics/apiseeds_lyrics.py:
```diff
@@ -0,0 +1,122 @@
+PLUGIN_DESCRIPTION = 'Fetch lyrics from Apiseeds Lyrics, which provides millions of lyrics from artist all around the world. ' \
+                     'Lyrics provided are for educational purposes and personal use only. Commercial use is not allowed. ' \
+                     'In order to use Apiseeds you need to get a free API key at <em>https://apiseeds.com</em>. ' \
+                     'Want to contribute? Check out the project page at <em>https://github.com/avalloneandrea/apiseeds-lyrics</em>!'''
```
I think an empty string `''` sneaked in at the end of this line ;)

phw reviewed (changes requested):
Looks good so far. Only the `error` needs to be handled and some minor tweak mainly to make codacy happy

avalloneandrea committed:
Fix method could be a function

avalloneandrea committed:
Small fixes

avalloneandrea committed:
Set apiseeds_apikey default value

avalloneandrea committed:
URL encode artist and title

avalloneandrea committed:
Set exceptions to catch

avalloneandrea committed:
Version 1.0.3

phw reviewed (approved):
This is looking good now to me, great work. I'll merge this and it will be available once we update the website (which hopefully will be soon, we are working on some website fixes currently).
